modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockStorage.scala

```diff
@@ -22,199 +22,116 @@ import org.tessellation.security.signature.Signed
 import io.chrisdavenport.mapref.MapRef
-import io.estatico.newtype.ops._
+import io.estatico.newtype.ops.toCoercibleIdOps
 import monocle.macros.syntax.lens._
 
-class BlockStorage[F[_]: Sync](blocks: MapRef[F, ProofsHash, Option[StoredBlock]], tipsConfig: TipsConfig) {
+class BlockStorage[F[_]: Sync: Random](blocks: MapRef[F, ProofsHash, Option[StoredBlock]]) {
 
   def areParentsAccepted(block: DAGBlock): F[Map[BlockReference, Boolean]] =
     block.parent.traverse { ref =>
       isBlockAccepted(ref).map(ref -> _)
     }.map(_.toList.toMap)
 
   private[block] def accept(hashedBlock: Hashed[DAGBlock]): F[Unit] =
-    blocks(hashedBlock.proofsHash)
-      .modify[Either[BlockAcceptanceError, Unit]] {
-        case Some(storedBlock) =>
-          storedBlock match {
-            case unknown: UnknownBlock => (unknown.some, BlockIsUnknown(hashedBlock.proofsHash).asLeft)
-            case WaitingBlock(_, usages) =>
-              (AcceptedBlock(hashedBlock, Accepted, usages, isTip = false).some, ().asRight)
-            case accepted: AcceptedBlock => (accepted.some, BlockAlreadyAccepted(hashedBlock.proofsHash).asLeft)
-          }
-        case None =>
-          (None, BlockDoesNotExist(hashedBlock.proofsHash).asLeft)
-      }
-      .flatMap(_.liftTo[F])
+    blocks(hashedBlock.proofsHash).modify {
+      case Some(storedBlock) =>
+        storedBlock match {
+          case WaitingBlock(_) =>
+            (AcceptedBlock(hashedBlock).some, ().asRight)
+          case accepted: AcceptedBlock =>
+            (accepted.some, BlockAcceptanceError(hashedBlock.proofsHash, "Accepted").asLeft)
+          case majority: MajorityBlock =>
+            (majority.some, BlockAcceptanceError(hashedBlock.proofsHash, "Majority").asLeft)
+        }
+      case None =>
+        (None, BlockAcceptanceError(hashedBlock.proofsHash, "Block not found!").asLeft)
+    }.flatMap(_.liftTo[F])
+      .flatMap(_ => updateParentUsages(hashedBlock))
 
   def store(hashedBlock: Hashed[DAGBlock]): F[Unit] =
-    blocks(hashedBlock.proofsHash)
-      .modify[Either[BlockStoringError, Unit]] {
-        case Some(storedBlock) =>
-          storedBlock match {
-            case UnknownBlock(usages) => (WaitingBlock(hashedBlock.signed, usages).some, ().asRight)
-            case waiting: WaitingBlock =>
-              (waiting.some, BlockAlreadyStoredAndWaiting(hashedBlock.proofsHash).asLeft)
-            case accepted: AcceptedBlock =>
-              (accepted.some, BlockAlreadyStoredAndAccepted(hashedBlock.proofsHash).asLeft)
-          }
-        case None => (WaitingBlock(hashedBlock.signed, Set.empty[ProofsHash]).some, ().asRight)
-      }
-      .flatMap(_.liftTo[F]) >> updateParentUsages(hashedBlock)
+    blocks(hashedBlock.proofsHash).modify {
+      case Some(storedBlock) =>
+        storedBlock match {
+          case waiting: WaitingBlock =>
+            (waiting.some, BlockAlreadyStoredError(hashedBlock.proofsHash, "Waiting").asLeft)
```

Comment:
Parametrized errors are fine but I do not like the fact that we manually put strings there. Especially that there is a fixed set of block types `String` kinda says that there are more types which may fit that error.

How about passing the classname instead?

```scala
case block @ BlockAlreadyStoredError(hashedBlock.proofsHash, block.getClass.getSimpleName).asLeft
```

Reply:
Already solved with passing current state and implicit show method defined in the scope.